### PR TITLE
Fix context-boundary max_new_tokens clipping

### PR DESCRIPTION
**Pull request by shz13009**
## TL;DR

- **Bug:** requests accepted at the advertised context boundary can return two fewer tokens than requested.
- **Cause:** a legacy output-token reserve introduced by #1748 remains in the scheduler even though exact-boundary requests are now valid (#22546) and page-aware KV admission is enforced separately (#25126).
- **Fix:** use `max_req_len - input_len + 1`, while leaving the separate page-rounded global KV-admission bound unchanged.
- **Safety:** generating `N` tokens forwards the prompt and only the first `N - 1` generated tokens; the final token is sampled and returned without another decode step.
- **Validation:** focused scheduler tests, end-to-end boundary tests on Qwen3/Qwen3.5, speculative-decoding coverage, and an independent behavior cross-check against Mini-SGLang.

## Motivation

Requests that fit the advertised context length are silently clipped by two
tokens in the scheduler. With `--context-length 4096`, `ignore_eos=True`, and
`max_new_tokens=16`:

```text
input=4078, requested=16, actual=16
input=4079, requested=16, actual=15
input=4080, requested=16, actual=14
input=4081, requested=16, HTTP 400
```

This reproduces on both Qwen3-0.6B and the hybrid GDN/linear-attention
Qwen3.5-0.8B path, so it is not model-specific.

SGLang's public validation permits requests whose total length is exactly the
model context length:

```text
input_len + max_new_tokens <= context_len
```

This boundary was explicitly adopted in #22546 after the same behavior was
reported in discussion #16020.

The scheduler uses a different internal length. In
[`TPModelWorker.get_worker_info`](https://github.com/sgl-project/sglang/blob/604d3561b012816aa4cec2fb89c522f819f68af8/python/sglang/srt/managers/tp_worker.py#L514-L519),
`max_req_len` is already defined as:

```python
max_req_len = min(
    context_len - 1,
    effective_max_total_num_tokens * dcp_size - 1,
)
```

However, [`Scheduler.init_req_max_new_tokens`](https://github.com/sgl-project/sglang/blob/604d3561b012816aa4cec2fb89c522f819f68af8/python/sglang/srt/managers/scheduler.py#L2140-L2148) applies the second subtraction:

```python
max_req_len - input_len - 1
```

The two commit-pinned assignments are therefore:

1. [`max_req_len = context_len - 1`](https://github.com/sgl-project/sglang/blob/604d3561b012816aa4cec2fb89c522f819f68af8/python/sglang/srt/managers/tp_worker.py#L513-L519)
2. [`max_new_tokens <= max_req_len - input_len - 1`](https://github.com/sgl-project/sglang/blob/604d3561b012816aa4cec2fb89c522f819f68af8/python/sglang/srt/managers/scheduler.py#L2140-L2148)

When the context-length bound is active, composing them gives
`max_new_tokens <= context_len - input_len - 2`, explaining the two-token clipping. The second subtraction was originally an intentional conservative reserve from #1748; the historical context below explains why it is redundant under the current contract and page-aware admission logic.

The execution invariant is that generating `N` output tokens forwards the
prompt and only the first `N - 1` generated tokens. The final token is sampled
and returned, but is never forwarded in another decode step. Therefore:

```text
input_len + max_new_tokens - 1 <= max_req_len
max_new_tokens <= max_req_len - input_len + 1
```

EOS does not require a separate reserved slot: when sampled, it is itself one
of the generated tokens. The page-rounded global KV admission bound remains a
separate safety constraint.

## Modifications

- Use `max_req_len - input_len + 1` for the per-request context bound.
- Keep the page-rounded global KV admission bound added by #25126 unchanged.
- Update the existing scheduler property tests to encode the forward-length
  invariant.
- Add a focused regression for a request exactly at the public context
  boundary. The regression fails on the parent commit with `14 != 16` and
  passes with this change.

After the change:

```text
input=4078, requested=16, actual=16
input=4079, requested=16, actual=16
input=4080, requested=16, actual=16
input=4081, requested=16, HTTP 400
```

## Historical Context

The extra scheduler subtraction was intentional when introduced by [#1748](https://github.com/sgl-project/sglang/pull/1748) in October 2024. That PR described three conservative rules: leave one token when deriving `max_req_len` from capacity, leave five tokens for `max_req_input_len`, and leave one more token when adjusting `max_new_tokens`. At that time, the goal was prefill-OOM robustness, and admission used a simpler strict estimate based on `extend_input_len + max_new_tokens < rem_total_tokens`.

Two later changes altered the contract and made the relevant safety check explicit:

- [#22546](https://github.com/sgl-project/sglang/pull/22546) changed public validation to allow `input_len + max_new_tokens == context_len`. It was approved by the maintainer who introduced #1748.
- [#25126](https://github.com/sgl-project/sglang/pull/25126) added the separate page-aware global KV bound `ceil_page(input_len) + max_new_tokens + page_size < max_total_num_tokens`, preventing requests that can never pass `PrefillAdder` admission from entering the waiting queue.

The historical reserve therefore should not be treated as an accidental typo. Under the current contract, however, retaining it in the context bound silently clips valid requests even when the newer explicit KV-admission bound has sufficient capacity. This PR removes that now-redundant context reserve while preserving the page-aware KV safety constraint.

## Follow-up

The `+1` conversion is necessary because `max_req_len` currently represents the maximum forwarded/KV-backed length, while the final sampled token is returned without another forward. A separate refactor could make this distinction explicit with names such as `max_forward_len` and `max_total_sequence_len`, then audit all consumers in the scheduler, attention metadata, speculative decoding, disaggregation, and memory pools. That broader semantic rename is intentionally out of scope here; this PR keeps the existing `max_req_len` contract and applies the localized conversion required to fix the boundary behavior.

## Mini-SGLang Cross-Check

Mini-SGLang independently implements the same public context-boundary behavior. Its scheduler computes:

```python
max_output_len = max_seq_len - input_len
```

Thus, for model context length `C`, input length `I`, and requested output length `O`, Mini-SGLang permits `O <= C - I`. See [Mini-SGLang scheduler intake](https://github.com/sgl-project/mini-sglang/blob/9a91cfafe754aa85daee49998176275667eb58f2/python/minisgl/scheduler/scheduler.py#L177-L188).

This is consistent with its execution state. A request sets `max_device_len = input_len + output_len`, and each forward increments `device_len` before the next token is sampled. Consequently, generating `O` output tokens forwards the prompt plus only the first `O - 1` generated tokens; the final sampled token is returned without another forward. See [request state](https://github.com/sgl-project/mini-sglang/blob/9a91cfafe754aa85daee49998176275667eb58f2/python/minisgl/core.py#L38-L58) and [engine forward](https://github.com/sgl-project/mini-sglang/blob/9a91cfafe754aa85daee49998176275667eb58f2/python/minisgl/engine/engine.py#L190-L207).

With core SGLang's `max_req_len = C - 1`, this PR's bound reduces to:

```text
O <= max_req_len - I + 1
  <= (C - 1) - I + 1
  <= C - I
```

The patched core context bound therefore matches Mini-SGLang's behavior. This is independent corroboration of the forward-length invariant, not a substitute for core-specific tests covering page-rounded global KV admission, DCP, speculative-decoding reservations, and other production modes.

## Accuracy Tests

This changes only scheduler length accounting, not model forward computation,
logits, or sampling math.

End-to-end validation used Qwen/Qwen3.5-0.8B on an RTX 3070 with the Triton
attention backend. All accepted requests through total length 4096 returned the
requested token count; total length 4097 remained rejected with HTTP 400.

The exact boundary also passed with Qwen3.5 NEXTN/EAGLE speculative decoding,
including its four reserved token slots; one token beyond it returned HTTP 400
and the server remained healthy.

Focused unit results:

```text
9 passed, 542 subtests passed
```

## Speed Tests and Profiling

No hot inference path changes. The patch replaces one integer expression during
request initialization, so no speed impact is expected.

Local `pre-commit run --all-files` passed all Python formatting, lint,
registered-test, and repository checks. Rust-only hooks could not run because
this WSL environment does not have `cargo` or `rustup`; this PR does not modify
Rust files.

No user-facing documentation change is required; the scheduler comment now
documents the corrected forward-length invariant.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: [Run #33008715565](https://github.com/sgl-project/sglang/actions/runs/33008715565)
Latest PR Test (Extra): :x: [Run #33008715247](https://github.com/sgl-project/sglang/actions/runs/33008715247)
Latest PR Test (AMD ROCm 7.2): :x: [Run #33008715424](https://github.com/sgl-project/sglang/actions/runs/33008715424)

### CODE REVIEW

**Comment by shz13009** [comment]
@merrymercy @hnyls2002 @cctry Could a Scheduler Merge Oncall please review this fix and trigger the appropriate CI when ready? The focused regression was verified to fail on the parent commit with 14 != 16 and pass with the fix; the full scheduler unit file passes with 8 tests and 542 property subtests.
